Uncertainty propagation library handling

phillipstanleymarbell opened the pull request:
Uncertainty propagation library handling.

harrysarson committed:
Use custom binutils

Rather than downloading a tar file, use git to pull down a tagged version.
Then apply a patch.

harrysarson committed:
update binutils patch

harrysarson committed:
Get loads and stores working

harrysarson committed:
TRANSITIVE CHANGES TO CONFIG

harrysarson committed:
Update binutil patch

harrysarson committed:
Add python scripts from madgwick benchmark

harrysarson committed:
Configure path to utilities

harrysarson committed:
improve error handling of fix-binary

harrysarson committed:
Update files in place

harrysarson committed:
add libuncertain

harrysarson committed:
update printf to version 4

SamuelmsWong-zz commented on conf/setup.conf:
```diff
@@ -1,23 +1,23 @@
 ##	You will want to change the following to suit your setup:
 ##
-SUNFLOWERROOT	= /tmp/sunflower-toolchain-github-clone
+SUNFLOWERROOT	= /home/harry/sunflower-simulator
```
Change back to default commenting.

harrysarson replied:
There is this nifty thing you can do on github reviews where you suggest changes by pressing `ctrl-g`.


```suggestion
SUNFLOWERROOT	= /tmp/sunflower-toolchain-github-clone
```

harrysarson replied:
Saying that, because @phillipstanleymarbell opened this pull request, I do not have the power to merge these suggestions via github. Will do in a push to this branch instead.

SamuelmsWong-zz replied:
Ah thanks nice to know.

SamuelmsWong-zz commented on conf/setup.conf:
```diff
@@ -1,23 +1,23 @@
 #       Specifies the architecture used. Uncomment for riscv builds.
 #
-#ADDITIONAL_ARCH_FLAGS = --with-cpu=rv32ifd
+ADDITIONAL_ARCH_FLAGS = --with-arch=rv32ifd
```
Use `with-arch`. Leave this as a comment though.

harrysarson replied:
```suggestion
#ADDITIONAL_ARCH_FLAGS = --with-arch=rv32ifd
```

harrysarson commented on tools/Makefile:
```diff
@@ -21,11 +22,11 @@ binutils-pre:
-	mv $(PREFIX)/bin/$(TARGET-ARCH)* $(TOOLS)/bin/;\
-	$(DEL) $(BINUTILS)/objdir;\
+	cp $(PREFIX)/bin/$(TARGET-ARCH)* $(TOOLS)/bin/;\
+	# $(DEL) $(BINUTILS)/objdir;\
```
```suggestion
	$(DEL) $(BINUTILS)/objdir;\
```

harrysarson commented:
Hi @SamuelmsWong thanks for you work on this!

I had a look at this branch and there is some bad new: it is not in a fit state to merge into master as is.

I patched the gnu binutils dependancy. Not to downplay my efforts here but the patch was a) ugly, b) possibly buggy c) specific to the uncertain work I was doing. I would not advocate building the assembler in this patched way for anything other than work on the uncertain framework. To merge this into master would cause every build of sunflower to use my patched version of binutils.

Saying that, the uncertain work I did will not work using a standard (unpatched) assembler so some compromise will be needed - possibly a side by side install.

The files that need a bit of love are:

* tools/Makefile
* tools/source/downloads.sh
* tools/source/patches/binutils/uncertain.patch

harrysarson committed:
apply review suggestions

harrysarson commented:
@SamuelmsWong I have given you push access to my fork - seems like the best way for you to work on this pull request :)

SamuelmsWong commented:
>To merge this into master would cause every build of sunflower to use my patched version of binutils.

@harrysarson I was having a think about how the assembly code is altered to use uncertain instrs (`(un)fadd.s`) and then assembled by the patched `binutils`. The new `binutils` patch does not affect old opcodes and only adds new opcodes.
Therefore it seems that as long as `.py` scripts are run only on the desired `.s` assembly files, other sunflower builds will not be affected by the patched `binutils`, because it will assemble floating point instrs as usual. This means that the patched `binutils` will act just the same despite having additional uncertain opcodes. Is this a correct understanding?

harrysarson commented:
> This means that the patched binutils will act just the same despite having additional uncertain opcodes. Is this a correct understanding?

In theory yes, however the theory depends on the quality of my patches and gnu binutils is a fairly garnly piece of software so I am far from 100% confident in my patches.

SamuelmsWong commented:
> In theory yes, however the theory depends on the quality of my patches and gnu binutils is a fairly garnly piece of software so I am far from 100% confident in my patches.

OK, thanks Harry. I'll have a read around and see what I can do.

samuelmswong committed:
removed buggy whitespace and completed d/dx of asin(x)

harrysarson commented on tools/source/uncertain/uncertain.S:
```diff
@@ -111,6 +111,7 @@ asinunf:
 	fmul.s    fa4, fs0, fs0
 	fsub.s    fa4, fa3, fa4
 	fsqrt.s   fa4, fa4
+	fdiv.s    fa4, fa3, fa4
```
Bah! Assembly is so hard to get correct. Nice catch.

SamuelmsWong commented:
Hi Harry, I'm kinda stuck. I've cloned your sunflower-simulator repo and tried to build it. The `.patch` file is run successfully from `downloads.sh` but when the toolchain Makefile tries to assemble `uncertain.S`, it fails due to the new instructions (e.g. `ungcov.s`) producing an assembler error message "`unrecognized opcode`". This is strange as you have added these already with the `.patch` file.

To get an understanding of which files are relevant, I tried:
- altering the error string in `gas/config/tc-riscv.c` from "`unrecognized opcode`" to "`unrecognized ooopcode`". After rebuilding (running `make binutils` and then `make uncertain`) there was no change.
- commenting out all the `ret` entries in `gas/opcodes/riscv-opc.c` and rebuilding to intentionally cause similar error - there was no change.
Is there something else I have to build as well in order to update the assembler, or am I working with the wrong files?

harrysarson commented:
Ah! I am very sorry. I know what I did wrong.

When building modified binutils I could not get it to work with existing make files (never found out why). Therefore, I commented out https://github.com/phillipstanleymarbell/Sunflower-toolchain/blob/c90bce0c0db0b78c7fad24191f75c7e16945551b/tools/Makefile#L25 this line.

Instead I build binutils by 
1. running `make binutils-pre` (not that this does not do making because of the comment).
2. `cd`ing into the `objdir` directory created by (1.) and running `make all install`
3. running `make binutils-post` 

I had forgetten about this hack I did. Really sorry to not have flagged earlier.

SamuelmsWong commented:
Thanks, I've got it running now.

harrysarson commented:
Nice!

harrysarson committed:
Merge branch 'master' into uncertain

harrysarson commented on conf/setup.conf:
```diff
@@ -14,11 +14,6 @@ TARGET-ARCH-FLAGS = -DM32
-#       Specifies the architecture used. Uncomment for riscv builds.
-#
-#ADDITIONAL_ARCH_FLAGS = --with-arch=rv32ifd
-
```
In 7a05e29c5a77906fbbace4691dc12e5a8793b9c5 @phillipstanleymarbell removed these lines from  `tools/Makefile`. I assume that this PR should not add them back in?

SamuelmsWong-zz replied:
I think so.

harrysarson committed:
undo changes to config

harrysarson committed:
[wip] toolchain for uncertain framework

I use ?= in setup.conf which allows overloading.
Otherwise make cross-xxx from the root does not work.

harrysarson commented:
Okay, I am running a build locally of the latest version. If it runs ok then I think this is ready to merge.

There are a couple of minor changes but I don't want to touch any files whilst the build is in progress.

With these changes, both normal binutils and uncertain binutils should be installed side by side. The uncertain version of the programs will have the `uncertain-` prefix. For example, `uncertain-riscv32-elf-as` instead of `riscv32-elf-as`.

harrysarson committed:
tidy + remove transitive changes

harrysarson commented:
Okay, @SamuelmsWong I think I am happy with the state of this now. When you get some time would it be possible to have a look?

The major change I made to existing code in this pull request is to use `?=` instead of `=` in `conf/setup.conf`. `?=` will not overwrite variables if they already exist which makes the logic in the root `Makefile` work.

Everything else should only effect builds that explicitly opt into the uncertain framework.

Fixes #19

SamuelmsWong commented:
> The major change I made to existing code in this pull request is to use `?=` instead of `=` in `conf/setup.conf`. `?=` will not overwrite variables if they already exist which makes the logic in the root `Makefile` work.

Hi Harry, I've tried to build this and have come across an error when building the HelloWorld benchmark. It attempts to build `cc` and `as` and returns errors such as `init-riscv.S:7: Error: unknown pseudo-op: `.option'`, whereas I would have expected it to use `riscv-elf-cc` and `riscv-elf as` defined in `setup.conf`. I was wondering whether you experienced anything similar, and if the `?=` is related to the error.
Currently, it will be excluded in the upcoming merge.

SamuelmsWong-zz committed:
Merge branch 'master' into uncertain

SamuelmsWong-zz committed:
Update setup.conf

harrysarson commented:
Thanks @SamuelmsWong, excluding from the merge sounds best. :+1:
